apps/opik-backend/src/main/java/com/comet/opik/domain/ProxyService.java

```diff
@@ -0,0 +1,131 @@
+package com.comet.opik.domain;
+
+import com.comet.opik.api.ProviderApiKey;
+import com.comet.opik.api.ProviderApiKeyUpdate;
+import com.comet.opik.api.error.EntityAlreadyExistsException;
+import com.comet.opik.api.error.ErrorMessage;
+import com.comet.opik.infrastructure.EncryptionService;
+import com.comet.opik.infrastructure.auth.RequestContext;
+import com.google.inject.ImplementedBy;
+import jakarta.inject.Inject;
+import jakarta.inject.Provider;
+import jakarta.inject.Singleton;
+import jakarta.ws.rs.NotFoundException;
+import jakarta.ws.rs.core.Response;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.jdbi.v3.core.statement.UnableToExecuteStatementException;
+import ru.vyarus.guicey.jdbi3.tx.TransactionTemplate;
+
+import java.sql.SQLIntegrityConstraintViolationException;
+import java.util.List;
+import java.util.UUID;
+
+import static com.comet.opik.infrastructure.db.TransactionTemplateAsync.READ_ONLY;
+import static com.comet.opik.infrastructure.db.TransactionTemplateAsync.WRITE;
+
+@ImplementedBy(ProxyServiceImpl.class)
+public interface ProxyService {
+
+    ProviderApiKey get(UUID id);
+    ProviderApiKey saveApiKey(ProviderApiKey providerApiKey);
+    void updateApiKey(UUID id, ProviderApiKeyUpdate providerApiKeyUpdate);
+}
+
+@Slf4j
+@Singleton
+@RequiredArgsConstructor(onConstructor_ = @Inject)
+class ProxyServiceImpl implements ProxyService {
+
+    private static final String PROVIDER_API_KEY_ALREADY_EXISTS = "Api key for this provider already exists";
+    private final @NonNull Provider<RequestContext> requestContext;
+    private final @NonNull IdGenerator idGenerator;
+    private final @NonNull TransactionTemplate template;
+    private final @NonNull EncryptionService encryptionService;
+
+    @Override
+    public ProviderApiKey get(UUID id) {
+        String workspaceId = requestContext.get().getWorkspaceId();
+
+        log.info("Getting provider api key with id '{}', workspaceId '{}'", id, workspaceId);
+
+        var providerApiKey =  template.inTransaction(READ_ONLY, handle -> {
+
+            var repository = handle.attach(ProviderApiKeyDAO.class);
+
+            return repository.fetch(id, workspaceId).orElseThrow(this::createNotFoundError);
+        });
+        log.info("Got provider api key with id '{}', workspaceId '{}'", id, workspaceId);
+
+        return providerApiKey.toBuilder()
+                .apiKey(encryptionService.decrypt(providerApiKey.apiKey()))
+                .build();
+    }
+
+    @Override
+    public ProviderApiKey saveApiKey(@NonNull ProviderApiKey providerApiKey) {
+        UUID apiKeyId = idGenerator.generateId();
+        String userName = requestContext.get().getUserName();
+        String workspaceId = requestContext.get().getWorkspaceId();
+
+        var newProviderApiKey = providerApiKey.toBuilder()
+                .id(apiKeyId)
+                .apiKey(encryptionService.encrypt(providerApiKey.apiKey()))
```

Comment:
Instead of doing this here, we could implement it in the Deserializer to make sure the value is already encrypted at the controller level

Reply:
Adds additional complexity, not intuitive. Also since it's needed only for Deserializer and for Serializer we don't need it (see below) I would rather not do it, to avoid confusion.

Reply:
The point is to centralize it and avoid leaking. Also, it would actually simplify the CRUD

Reply:
I agree with Thiago here. The payload of the API key should be encrypted at the earliest stage possible, in order to protect from any leakage (logs, memory attacks etc.) A different story is that we compromise for earlier development. But it's concerning to compromise in relation to the custody of customer's external keys.

Reply:
It could be as  simple as that:

```java
@JsonView({View.Write.class}) @NotBlank @JsonDeserialize(using = ProviderApiKey.ProviderApiKeyDeserializer.class) String apiKey
```

```java
   public class ProviderApiKeyDeserializer extends JsonDeserializer<String> {

        @Override
        public String deserialize(JsonParser parser, DeserializationContext ctx) throws IOException {
            return EncryptionUtils.encrypt(parser.getText());
        }
    }
```

Reply:
I made corresponding update, but it's not that simple:) EncryptionUtils has to be static in order to be used here, since we can't inject anything to this custom deserializer.